### ISSUE TITLE
Khala nodes leading to spam on https://telemetry.polkadot.io 

### ISSUE THREAD

**Issue by jsdw** [Phala-Network/khala-parachain#19]
Hey guys,

It looks like nodes advertising themselves as "Khala Node v0.1.2-ae5d185(..)" are clogging up the telemetry page (https://telemetry.polkadot.io/#list/0xb0a8d493285c2df73290dfb7e61f870f17b41801197a149ca93654499ea3dafe), as they are sending telemetry with continually changing message and network IDs on the same persistent telemetry connection, leading to an endlessly growing list of "stale" nodes being displayed there.

I'm not sure where the issue lies, but perhaps it's worth redirecting the node telemetry elsewhere to avoid overwhelming the telemetry server until the issue has been resolved? :)

**Comment by jasl** [comment]
Yeah we got report of this, here's context on our side:

- We don't touch the codebase of telemetry
- The telemetry is sending by Polkadot client of a parachain node
- We're trying to figure out why there's so much 0# node, that's looks not normal

We'd glad to help, but it's difficult to "redirecting the node telemetry", at this time we can make a new Docker image that redirect `tetemetry.polkadot.io` to `localhost`, but if users won't upgrade their node...

BTW, is it possible to ban Khala node from `tetemetry.polkadot.io` side?

**Comment by jsdw** [comment]
Thanks for the info! On the telemetry side, we are going to deploy some node-name deduping shortly (and possibly just temporarily) that should ease things up. We could do something more specific if need be, but hopefully deduping by node name on a given connection will do the trick for now!

**Comment by jasl** [comment]
> Thanks for the info! On the telemetry side, we are going to deploy some node-name deduping shortly (and possibly just temporarily) that should ease things up. We could do something more specific if need be, but hopefully deduping by node name on a given connection will do the trick for now!

Thank you!

I'm also thinking some improvements:

- A Parachain node (no mattter Khala, Acala or something else) is also a Kusama node, it's no need to send telemetry as Kusama, that's duplicate report
- A node stuck on 0# block height isn't normal, maybe tetemetry service should ignore them

**Comment by jsdw** [comment]
The 0 block height reported in the telemetry UI might just be because that's the default value reported when we see a "system connected" message come in. I'm guessing that no messages arrive after that to update the known node state, and so it is just stuck at reporting "0".

**Comment by maciejhirsz** [comment]
So it looks like the duplicate nodes are starting separate connections, hence my deduping attempts weren't fruitful so far. I'm trying to run a Khala Node locally so I can send telemetry towards a local backend, but I'm running into issues:

```
$ yarn start ./khala.config.json
yarn run v1.22.10
$ yarn build && node dist/cli.js ./khala.config.json
$ tsc

🧹 Resolving parachain id...
2021-09-21 09:02:26 Loading spec: khala-local, custom parachain-id = "2004"    
2021-09-21 09:02:26 Building chain spec    
  ✓ Read parachain id for ./bin/khala-node: 2004
`reuseChainSpec` flag enabled, will use existing `rococo-local-raw.chain_spec.json`, delete it if you don't want to reuse
Starting alice...
Added --node-key=807e26fb3fd3d694b8b5f78a04d4bcff2e94a6f2e87bc809d9690e771bba8aca,--rpc-cors=all,--telemetry-url='ws://127.0.0.1:8000/submit 0'
Full args "--chain=/home/maciej/rust/khala-parachain/polkadot-launch/rococo-local-raw.chain_spec.json,--ws-port=19944,--port=10333,--alice,--base-path=data/relay_alice,--node-key=807e26fb3fd3d694b8b5f78a04d4bcff2e94a6f2e87bc809d9690e771bba8aca,--rpc-cors=all,--telemetry-url='ws://127.0.0.1:8000/submit 0'"
Starting bob...
Full args "--chain=/home/maciej/rust/khala-parachain/polkadot-launch/rococo-local-raw.chain_spec.json,--ws-port=19244,--port=12333,--bob,--base-path=data/relay_bob"
Starting charlie...
Full args "--chain=/home/maciej/rust/khala-parachain/polkadot-launch/rococo-local-raw.chain_spec.json,--ws-port=19344,--port=13333,--charlie,--base-path=data/relay_charlie"
Starting dave...
Full args "--chain=/home/maciej/rust/khala-parachain/polkadot-launch/rococo-local-raw.chain_spec.json,--ws-port=19444,--port=14333,--dave,--base-path=data/relay_dave"
2021-09-21 09:02:26          API-WS: disconnected from ws://127.0.0.1:19944: 1006:: connection failed
2021-09-21 09:02:29          API-WS: disconnected from ws://127.0.0.1:19944: 1006:: connection failed
2021-09-21 09:02:31          API-WS: disconnected from ws://127.0.0.1:19944: 1006:: connection failed
2021-09-21 09:02:34          API-WS: disconnected from ws://127.0.0.1:19944: 1006:: connection failed
```

Getting the same issue on `_dev` and `_local` configs.

**Comment by jasl** [comment]
> So it looks like the duplicate nodes are starting separate connections, hence my deduping attempts weren't fruitful so far. I'm trying to run a Khala Node locally so I can send telemetry towards a local backend, but I'm running into issues:
> 
> ```
> $ yarn start ./khala.config.json
> yarn run v1.22.10
> $ yarn build && node dist/cli.js ./khala.config.json
> $ tsc
> 
> 🧹 Resolving parachain id...
> 2021-09-21 09:02:26 Loading spec: khala-local, custom parachain-id = "2004"    
> 2021-09-21 09:02:26 Building chain spec    
>   ✓ Read parachain id for ./bin/khala-node: 2004
> `reuseChainSpec` flag enabled, will use existing `rococo-local-raw.chain_spec.json`, delete it if you don't want to reuse
> Starting alice...
> Added --node-key=807e26fb3fd3d694b8b5f78a04d4bcff2e94a6f2e87bc809d9690e771bba8aca,--rpc-cors=all,--telemetry-url='ws://127.0.0.1:8000/submit 0'
> Full args "--chain=/home/maciej/rust/khala-parachain/polkadot-launch/rococo-local-raw.chain_spec.json,--ws-port=19944,--port=10333,--alice,--base-path=data/relay_alice,--node-key=807e26fb3fd3d694b8b5f78a04d4bcff2e94a6f2e87bc809d9690e771bba8aca,--rpc-cors=all,--telemetry-url='ws://127.0.0.1:8000/submit 0'"
> Starting bob...
> Full args "--chain=/home/maciej/rust/khala-parachain/polkadot-launch/rococo-local-raw.chain_spec.json,--ws-port=19244,--port=12333,--bob,--base-path=data/relay_bob"
> Starting charlie...
> Full args "--chain=/home/maciej/rust/khala-parachain/polkadot-launch/rococo-local-raw.chain_spec.json,--ws-port=19344,--port=13333,--charlie,--base-path=data/relay_charlie"
> Starting dave...
> Full args "--chain=/home/maciej/rust/khala-parachain/polkadot-launch/rococo-local-raw.chain_spec.json,--ws-port=19444,--port=14333,--dave,--base-path=data/relay_dave"
> 2021-09-21 09:02:26          API-WS: disconnected from ws://127.0.0.1:19944: 1006:: connection failed
> 2021-09-21 09:02:29          API-WS: disconnected from ws://127.0.0.1:19944: 1006:: connection failed
> 2021-09-21 09:02:31          API-WS: disconnected from ws://127.0.0.1:19944: 1006:: connection failed
> 2021-09-21 09:02:34          API-WS: disconnected from ws://127.0.0.1:19944: 1006:: connection failed
> ```
> 
> Getting the same issue on `_dev` and `_local` configs.

`local` and `dev` should pair with `rococo-local`, I guess, the node will report to telemetry service twice, one is for `khala` another one is for `rococo`.

I have a question,  if a node start and close to frequently (some miners are not experienced ops, they only know stop & start when they meet problems) does the telemetry service can handle this case well?

**Comment by maciejhirsz** [comment]
> I have a question, if a node start and close to frequently (some miners are not experienced ops, they only know stop & start when they meet problems) does the telemetry service can handle this case well?

It should in principle, we want to investigate what's happening there exactly because it looks like stuff isn't being cleared properly.

**Comment by jsdw** [comment]
Is there a chance that the issues you're encountering starting a node are currently also affecting Khala nodes, leading the mto be restarted continuously, which leads to a load of nodes on telemetry owing to us not cleaning up properly or something?

**Comment by maciejhirsz** [comment]
@jasl can you do a quick start up guide for this, because following README gets me nowhere. I can start a rococo node separately from the yarn script, but it doesn't change anything there. It does start 3-4 polkadot processes on my machine, but none of them seem to connect to my local telemetry, though not sure if that's because of the error in the log, or the args aren't being passed in full.

**Comment by jasl** [comment]
> @jasl can you do a quick start up guide for this, because following README gets me nowhere. I can start a rococo node separately from the yarn script, but it doesn't change anything there. It does start 3-4 polkadot processes on my machine, but none of them seem to connect to my local telemetry, though not sure if that's because of the error in the log, or the args aren't being passed in full.

You wanna add `--telemetry-url='ws://127.0.0.1:8000/submit 0'` arg to relaychain and parachain right?

**Comment by maciejhirsz** [comment]
Yup

**Comment by jasl** [comment]
> Yup

I add `--telemetry-url='ws://127.0.0.1:8000/submit 0'` then got error
`error: Invalid value for '--telemetry-url <URL VERBOSITY>...': invalid digit found in string`

**Comment by jasl** [comment]
You can edit `khala_dev.config.json` or create a new config for example `local.config.json`

The content looks like

```json
{
  "relaychain": {
    "bin": "./bin/polkadot",
    "chain": "rococo-local",
    "nodes": [
      {
        "name": "alice",
        "wsPort": 19944,
        "port": 10333,
        "flags": [
          "--node-key=807e26fb3fd3d694b8b5f78a04d4bcff2e94a6f2e87bc809d9690e771bba8aca",
          "--rpc-cors=all",
          "--telemetry-url='ws://127.0.0.1:8000/submit 0'"
        ]
      },
      {
        "name": "bob",
        "wsPort": 19244,
        "port": 12333,
        "flags": [
          "--telemetry-url='ws://127.0.0.1:8000/submit 0'"
        ]
      },
      {
        "name": "charlie",
        "wsPort": 19344,
        "port": 13333,
        "flags": [
          "--telemetry-url='ws://127.0.0.1:8000/submit 0'"
        ]
      },
      {
        "name": "dave",
        "wsPort": 19444,
        "port": 14333,
        "flags": [
          "--telemetry-url='ws://127.0.0.1:8000/submit 0'"
        ]
      }
    ],
    "genesis": {
      "runtime": {
        "runtime_genesis_config": {
          "configuration": {
            "config": {
              "validation_upgrade_frequency": 1,
              "validation_upgrade_delay": 1
            }
          }
        }
      }
    }
  },
  "parachains": [
    {
      "bin": "./bin/khala-node",
      "chain": "khala-dev-2004",
      "balance": "1000000000000000000000",
      "nodes": [
        {
          "name": "alice",
          "wsPort": 9944,
          "port": 30333,
          "flags": [
            "--node-key=81bbf2ee046d4f57ab9ce02e822ac2a3192e4a7b38af757383f3dcb039c15e05",
            "--ws-external",
            "--rpc-external",
            "--rpc-cors=all",
            "--rpc-methods=Unsafe",
            "--telemetry-url='ws://127.0.0.1:8000/submit 0'",
            "--",
            "--node-key=1335ac151f2d804189bfc68bd0843a676a3dc2aa6a208326970a75e6c75d4226",
            "--port=30334",
            "--ws-port=9945",
            "--pruning=archive",
            "--ws-external",
            "--rpc-external",
            "--rpc-cors=all",
            "--telemetry-url='ws://127.0.0.1:8000/submit 0'"
          ]
        },
        {
          "name": "bob",
          "wsPort": 9244,
          "port": 32333,
          "flags": [
            "--telemetry-url='ws://127.0.0.1:8000/submit 0'",
            "--",
            "--telemetry-url='ws://127.0.0.1:8000/submit 0'"
          ]
        }
      ]
    }
  ],
  "simpleParachains": [],
  "hrmpChannels": [],
  "types": {},
  "finalization": false
}
```

It should work after we resolve `error: Invalid value for '--telemetry-url <URL VERBOSITY>...': invalid digit found in string`

**Comment by maciejhirsz** [comment]
@jasl doesn't seem to work for me, I also get this:

```
2021-09-21 13:24:59        API/INIT: Error: FATAL: Unable to initialize the API: Cannot convert metadata from version 14 to 13
    at EventEmitter.value (/home/maciej/rust/khala-parachain/polkadot-launch/node_modules/@polkadot/api/base/Init.cjs:82:25)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

Might be because my polkadot binary is built from latest master.

**Comment by jasl** [comment]
> @jasl doesn't seem to work for me, I also get this:
> 
> ```
> 2021-09-21 13:24:59        API/INIT: Error: FATAL: Unable to initialize the API: Cannot convert metadata from version 14 to 13
>     at EventEmitter.value (/home/maciej/rust/khala-parachain/polkadot-launch/node_modules/@polkadot/api/base/Init.cjs:82:25)
>     at processTicksAndRejections (internal/process/task_queues.js:95:5)
> ```
> 
> Might be because my polkadot binary is built from latest master.

Oh, try `polkadot-0.9.9` for now, Polkadot master introduce MetadataV14, I believe `polkadot.js` need to do adaption

**Comment by maciejhirsz** [comment]
Ok, that fixed the metadata error but still no telemetry. Is there a CLI flag to see the logs from the nodes themselves?

**Comment by maciejhirsz** [comment]
Got it, `"--telemetry-url=ws://127.0.0.1:8000/submit 0"` works, quotes are superficial since args are already separated when passed into the binary.

**Comment by jasl** [comment]
> Ok, that fixed the metadata error but still no telemetry. Is there a CLI flag to see the logs from the nodes themselves?

You can `ls` at `polkadot-launch` folder, you can see many `.log` files, `alice.log`, `bob.log`, `charlie.log`, etc. are Relaychains, and `9944.log`, `9144.log`, `9244.log`, etc. are Parachains.

**Comment by jasl** [comment]
> Got it, `"--telemetry-url=ws://127.0.0.1:8000/submit 0"` works, quotes are superficial since args are already separated when passed into the binary.

I see log shows 

`2021-09-21 20:43:30 ❌ Error while dialing /ip4/127.0.0.1/tcp/8000/x-parity-ws/%2Fsubmit: Custom { kind: Other, error: Other(B(A(Handshake(Io(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }))))) }`

looks it works.

**Comment by maciejhirsz** [comment]
There is no need to modify anything in the chain spec, telemetry settings are CLI only. I tried `"chain": "kusama"` and `"reuseChainSpec": true`, but it shuts down immediately:

```
$ yarn start khala_local.config.json 
yarn run v1.22.10
$ yarn build && node dist/cli.js khala_local.config.json
$ tsc

🧹 Resolving parachain id...
2021-09-21 14:08:15 Loading spec: khala-local, custom parachain-id = "2004"    
2021-09-21 14:08:15 Building chain spec    
  ✓ Read parachain id for ./bin/khala-node: 2004
2021-09-21 14:08:15 Building chain spec    

Killing all processes...
Done in 4.57s.
```

**Comment by jasl** [comment]
> There is no need to modify anything in the chain spec, telemetry settings are CLI only. I tried `"chain": "kusama"` and `"reuseChainSpec": true`, but it shuts down immediately:
> 
> ```
> $ yarn start khala_local.config.json 
> yarn run v1.22.10
> $ yarn build && node dist/cli.js khala_local.config.json
> $ tsc
> 
> 🧹 Resolving parachain id...
> 2021-09-21 14:08:15 Loading spec: khala-local, custom parachain-id = "2004"    
> 2021-09-21 14:08:15 Building chain spec    
>   ✓ Read parachain id for ./bin/khala-node: 2004
> 2021-09-21 14:08:15 Building chain spec    
> 
> Killing all processes...
> Done in 4.57s.
> ```

That because build chain spec stage failed, and polkadot-launch doesn't show any logs on this stage...

**Comment by maciejhirsz** [comment]
How do you guys run your live Kusama nodes then? :)

Also, while we are trying to reproduce this, can you add `--no-telemetry` args to your live nodes?

**Comment by jasl** [comment]
> How do you guys run your live Kusama nodes then? :)
> 
> Also, while we are trying to reproduce this, can you add `--no-telemetry` arg to your live nodes?

That's no problem for us, but, acutally, we only run ~20 nodes as bootnodes, others are our miners, if they won't upgrade their nodes, it can't help current situation...

From previous context, spam nodes report as `khala-node 0.1.2` but our latest version is `khala-node 0.1.4`

**Comment by maciejhirsz** [comment]
Let me rephrase that question, how can _I_ run a miner node that connects to Kusama?

**Comment by maciejhirsz** [comment]
Ok, cheers, I think I can work from there on.

**Comment by jsdw** [comment]
Out of interest, has something changed from your side @jasl in the last day or so? I'm no longer seeing an issue in Telemetry offhand! We did release a new version which has more protections in place, but I wouldn't have expected that to resolve the issue.

**Comment by jasl** [comment]
> Out of interest, has something changed from your side @jasl in the last day or so? I'm no longer seeing an issue in Telemetry offhand! We did release a new version which has more protections in place, but I wouldn't have expected that to resolve the issue.

No, I didn't do anything

**Comment by jsdw** [comment]
I'm happy to close this for now, as I haven't noticed an issue with Khala notes of late :)
